File: py/game.py
# -*- coding: utf-8 -*-
from copy import deepcopy
from random import randint
from threading import Thread
from time import time


action_space = [0, 0, 0, 0]
action_space[0] = [(i, j) for i in range(6)
                   for j in range(6) if i != j]
action_space[1] = [[act1, act2]
                   for act1 in action_space[0]
                   for act2 in action_space[0]]
action_space[2] = [[act1, act2, act3]
                   for act1 in action_space[0]
                   for act2 in action_space[0]
                   for act3 in action_space[0]]
action_space[3] = [[act1, act2, act3, act4]
                   for act1 in action_space[0]
                   for act2 in action_space[0]
                   for act3 in action_space[0]
                   for act4 in action_space[0]]

# print(action_space[0])


class MyThread(Thread):

    game = None
    acts = None
    k = 0
    rets = None
    nnn = 0
    last_inv = None

    def __init__(self, game, act, k):
        Thread.__init__(self)
        self.game = deepcopy(game)
        self.act = act
        self.k = k
        self.nnn = len(action_space[k-2])
        self.rets = [None for i in range(self.nnn)]

    def run(self):
        a = time()
        ret = self.game.step(self.act)
        if ret == "INVALID" or ret == "DEAD":
            self.rets = [ret for _ in range(self.nnn)]
            b = time()
            print(b-a)
            return

        # self.game.Drop()

        a = time()
        actions = action_space[self.k-2]
        i = 0
        while i < len(actions):
            g_ = deepcopy(self.game)
            acts = actions[i]
            for j in range(len(acts)):
                ret = g_.step(acts[j])
                if ret == "INVALID" or ret == "DEAD":
                    if len(acts)-j == 3:
                        self.rets[i:i+900] = [ret for _ in range(900)]
                        i += 900
                    elif len(acts)-j == 2:
                        self.rets[i:i+30] = [ret for _ in range(30)]
                        i += 30
                    elif len(acts)-j == 1:
                        self.rets[i] = ret
                        i += 1
                    break
            if ret == "INVALID" or ret == "DEAD":
                continue
            g_.value()
            self.rets[i] = g_
            i += 1

        b = time()
        print(b-a)
        return

    def get_results(self):
        return self.rets


class Game():

    m, n = 7, 6
    state = None
    drop = None
    score = 0
    v = 0

    def __init__(self):
        pass

    def reset(self):
        self.state = [[1, 1] for _ in range(self.n)]
        self.score = 0

    def move(self, act):
        fr, to = self.state[act[0]], self.state[act[1]]
        if len(fr) >= 2:
            for _ in range((fr[-1] == fr[-2])+1):
                to.append(fr.pop(-1))
        elif len(fr) == 1:
            to.append(fr.pop(-1))
        else:
            # print("Invalid Move.")
            return "INVALID"

    def comb(self, i):
        to = self.state[i]
        if len(to) >= 4:
            if to[-1] == to[-2] and to[-1] == to[-3] and to[-1] == to[-4]:
                to.pop(-1)
                to.pop(-1)
                to.pop(-1)
                to.append(to.pop(-1)+1)
            else:
                pass
        if len(to) >= 3:
            if to[-1] == to[-2] and to[-1] == to[-3]:
                to.pop(-1)
                to.pop(-1)
                to.append(to.pop(-1)+1)
        if len(to) >= 3:
            if to[-1] == to[-2] and to[-1] == to[-3]:
                to.pop(-1)
                to.pop(-1)
                to.append(to.pop(-1)+1)
        if len(to) >= 3:
            if to[-1] == to[-2] and to[-1] == to[-3]:
                to.pop(-1)
                to.pop(-1)
                to.append(to.pop(-1)+1)
        if to[-1] == 9:
            to.pop(-1)
            self.score += 1

    def isDead(self, i):
        to = self.state[i]
        if len(to) > self.m:
            # print("Game Over.")
            return "DEAD"

    def step(self, act):
        # move
        ret = self.move(act)
        if ret == "INVALID":
            return "INVALID"

        # Comb
        self.comb(act[1])

        # Dead
        ret = self.isDead(act[1])
        if ret == "DEAD":
            return "DEAD"

        return "VALID"

    def setDrop(self, c):
        self.drop = c

    def Drop(self):
        for i in range(self.n):
            self.state[i].append(self.drop[i])
            self.comb(i)
            ret = self.isDead(i)
            if ret == "DEAD":
                return "DEAD"
        return "Safe"

    def genDrop(self):
        mm = []
        for x in self.state:
            if len(x) > 0:
                mm.append(max(x))
            else:
                mm.append(0)
        mm = max(mm)
        drop = [randint(mm-1, mm+1) for _ in range(self.n)]
        for i in range(self.n):
            drop[i] = min(max(drop[i], 1), 7)
        return drop

    def print_state(self):
        tmp = [[0 for _ in range(self.n)] for _ in range(self.m)]
        for i in range(self.n):
            for j in range(len(self.state[i])):
                tmp[j][i] = self.state[i][j]
        print("Score: ", self.score)
        print("State: ")
        for i in range(self.m-1, -1, -1):
            if sum(tmp[i]) > 0:
                print(tmp[i])

    def value(self):
        # 剩余个数
        def f1(g):
            return sum([len(x) for x in g.state])

        # 剩余最高高度
        def f2(g):
            return max([len(x) for x in g.state])

        # 最大珠子数
        def f3(g):
            return max([max(x) for x in g.state if len(x) > 0])

        # 非空杆数
        def f4(g):
            return sum([len(x) != 0 for x in g.state])
        g = self
        self.v = f1(g), f2(g), f4(g)

    def solveK_v1(self, k):
        a = time()
        actions = action_space[k-1]
        rets = [0 for _ in range(len(actions))]
        for i in range(len(actions)):
            acts = actions[i]
            g_ = deepcopy(self)
            rets[i] = g_.sim(acts)
        b = time()
        print(b-a)
        best = None
        best_v = 0
        best_i = 0
        for i in range(len(rets)):
            ret = rets[i]
            if ret not in ["INVALID", "DEAD"]:
                if best is None or ret.v < best_v:
                    best = ret
                    best_v = best.v
                    best_i = i
        return actions[best_i], best

    def solveK_v2(self, k):
        actions = action_space[0]
        nnn = len(actions)
        ths = [None for _ in range(nnn)]
        rets = []
        th1 = [0, 1, 2, 3, 4, 5, 6, 7]
        th2 = [8, 9, 10, 11, 12, 13, 14, 15]
        th3 = [16, 17, 18, 19, 20, 21, 22, 23]
        th4 = [24, 25, 26, 27, 28, 29]
        for i in th1:
            ths[i] = MyThread(self, actions[i], k)
            ths[i].start()
        for i in th1:
            ths[i].join()
            rets.extend(ths[i].get_results())
        for i in th2:
            ths[i] = MyThread(self, actions[i], k)
            ths[i].start()
        for i in th2:
            ths[i].join()
            rets.extend(ths[i].get_results())
        for i in th3:
            ths[i] = MyThread(self, actions[i], k)
            ths[i].start()
        for i in th3:
            ths[i].join()
            rets.extend(ths[i].get_results())
        for i in th4:
            ths[i] = MyThread(self, actions[i], k)
            ths[i].start()
        for i in th4:
            ths[i].join()
            rets.extend(ths[i].get_results())

        best = None
        best_v = 0
        best_i = 0
        for i in range(len(rets)):
            ret = rets[i]
            if ret not in ["INVALID", "DEAD"]:
                if best is None or ret.v < best_v:
                    best = ret
                    best_v = best.v
                    best_i = i
        return action_space[k-1][best_i], best, rets

    def solveK(self, k):
        # return self.solveK_v1(k)
        a = time()
        ret2 = self.solveK_v1(k)
        b = time()
        print("total time: ", b-a)

        return ret2[0], ret2[1]

    def sim(self, acts):
        for act in acts:
            ret = self.step(act)
            if ret == "INVALID":
                return "INVALID"
            elif ret == "DEAD":
                return "DEAD"
        self.value()
        return self

File: cpp/Game.cpp
#include <atomic>
#include <cstdio>
#include <numeric>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <utility>
#include <vector>

using namespace std;

int score = 0;
typedef vector<vector<char>> State;
typedef vector<char> Array;
State state;
Array dropping;
typedef pair<char, char> Act;
vector<Act> Actions;
typedef vector<int> Value;
typedef pair<State, vector<Act>> RET;
#define INV -1
#define DEAD -2

void reset(State &state) {
  score = 0;
  state.resize(6);
  for (int i = 0; i < 6; i++)
    state[i] = vector<char>({1, 1});
  for (int i = 0; i < 6; i++)
    for (int j = 0; j < 6; j++)
      if (i != j)
        Actions.push_back(Act(i, j));
};

int sum(Array &t) {
  int s = 0;
  for (auto &i : t)
    s += i;
  return s;
}

int max(Array &t) {
  int s = 0;
  for (auto &i : t)
    if (i > s)
      s = i;
  return s;
}

int value_less(Value &va, Value &vb) {
  for (int i = 0; i < va.size(); i++)
    if (va[i] > vb[i])
      return 0;
  return 1;
};

void print_state(State &state) {
  State tmp(7, vector<char>(6, 0));
  for (int i = 0; i < 6; i++)
    for (int j = 0; j < state[i].size(); j++)
      tmp[j][i] = state[i][j];

  printf("Score: %d\n", score);
  for (int i = 6; i >= 0; i--) {
    if (sum(tmp[i]) > 0) {
      for (auto &j : tmp[i])
        printf("%d ", j);
      printf("\n");
    };
  }
};

State copy(State &state) {
  State ret;
  for (auto &t : state) {
    ret.push_back(Array(t));
  }
  return ret;
};

int move(State &state, Act &act) {
  int f = act.first, t = act.second;
  int n = state[act.first].size();
  if (n == 0)
    return INV;
  else if (n == 1) {
    state[t].push_back(state[f][n - 1]);
    state[f].pop_back();
  } else {
    if (state[f][n - 1] == state[f][n - 2]) {
      state[t].push_back(state[f][n - 1]);
      state[t].push_back(state[f][n - 1]);
      state[f].pop_back();
      state[f].pop_back();
    } else {
      state[t].push_back(state[f][n - 1]);
      state[f].pop_back();
    }
  };
  return 0;
};

void comb(Array &t) {
  int n = t.size();
  if (n >= 4) {
    if ((t[n - 1] == t[n - 2]) and (t[n - 1] == t[n - 3]) and
        (t[n - 1] == t[n - 4])) {
      t.pop_back();
      t.pop_back();
      t.pop_back();
      t.back() += 1;
    };
  };
  n = t.size();
  if (n >= 3)
    if ((t[n - 1] == t[n - 2]) and (t[n - 1] == t[n - 3])) {
      t.pop_back();
      t.pop_back();
      t.back() += 1;
      n = t.size();
      if ((t[n - 1] == t[n - 2]) and (t[n - 1] == t[n - 3])) {
        t.pop_back();
        t.pop_back();
        t.back() += 1;
        n = t.size();
        if ((t[n - 1] == t[n - 2]) and (t[n - 1] == t[n - 3])) {
          t.pop_back();
          t.pop_back();
          t.back() += 1;
        };
      };
    };
  if (t.back() == 9)
    t.pop_back();
};

int step(State &state, Act &act) {
  int ret = move(state, act);
  if (ret == INV)
    return INV;

  comb(state[act.second]);

  if (state[act.second].size() > 7)
    return DEAD;

  return 0;
};

void set_state(State &state, State &n_state) { state = n_state; }

int drop(State &state) {
  for (int i = 0; i < 6; i++) {
    state[i].push_back(dropping[i]);
    comb(state[i]);
    if (state[i].size() >= 7)
      return DEAD;
  }
  return 0;
}

Value value(State &state) {
  Array len;
  for (auto &t : state)
    len.push_back(t.size());
  Array empty;
  for (auto &t : state)
    len.push_back(t.size() != 0);
  Value v;
  v.push_back(sum(len));
  v.push_back(max(len));
  v.push_back(sum(empty));
  return v;
};

RET solveK1(State &state) {
  State best;
  vector<Act> best_acts;
  Value best_v = Value({0, 0, 0, 0});
  for (auto &a1 : Actions) {
    State t1 = State(state);
    int r1 = step(t1, a1);
    printf("action : %d %d\n", a1.first, a1.second);
    print_state(t1);
    if ((r1 == INV) or (r1 == DEAD)) {
      ;
    } else {
      Value v1 = value(t1);
      if (value_less(best_v, v1)) {
        best = t1;
        best_v = v1;
        best_acts = vector<Act>({a1});
      }
    }
  };
  return RET(best, best_acts);
};

RET solveK4_v1(State &state) {
  State best;
  vector<Act> best_acts;
  Value best_v = Value({100, 100, 100});
  for (auto &a1 : Actions) {
    State t1 = State(state);
    int r1 = step(t1, a1);
    if ((r1 == INV) or (r1 == DEAD)) {
      ;
    } else {
      for (auto &a2 : Actions) {
        State t2 = State(t1);
        drop(t2);
        int r2 = step(t2, a2);
        if ((r2 == INV) or (r2 == DEAD)) {
          ;
        } else {
          for (auto &a3 : Actions) {
            State t3 = State(t2);
            int r3 = step(t3, a3);
            if ((r3 == INV) or (r3 == DEAD)) {
              ;
            } else {
              for (auto &a4 : Actions) {
                State t4 = State(t3);
                int r4 = step(t4, a4);
                if ((r4 == INV) or (r4 == DEAD)) {
                  ;
                } else {
                  Value v4 = value(t4);
                  if (value_less(v4, best_v)) {
                    best = t4;
                    best_v = v4;
                    best_acts = vector<Act>({a1, a2, a3, a4});
                  }
                }
              };
            }
          };
        }
      };
    }
  };
  return RET(best, best_acts);
};

RET solveK(State &state, int k) {
  time_t begin_t = clock();
  RET ret;
  if (k == 4)
    ret = solveK4_v1(state);
  if (k == 1)
    ret = solveK1(state);
  time_t finish_t = clock();
  printf("\ntime: %fs\n", (double)(finish_t - begin_t) / CLOCKS_PER_SEC);
  return ret;
};

void test1(State &state) {
  reset(state);
  print_state(state);
  Act act = Act(1, 2);
  step(state, act);
  print_state(state);
};

void test2(State &state) {
  reset(state);
  // print_state(state);
  state = State({{}, {3}, {}, {}, {}, {}});
  // print_state(state);
  dropping = Array({4, 2, 3, 3, 2, 2});
  // drop(state);
  print_state(state);
  RET ret = solveK(state, 4);
  print_state(ret.first);

  dropping = Array({4, 4, 5, 5, 3, 4});
  state = State(ret.first);
  print_state(state);
  ret = solveK(state, 4);
  print_state(ret.first);
};

void test3(State &state) {
  reset(state);
  state = State({{5, 2, 5},
                 {5, 3, 7},
                 {4, 4, 6, 6},
                 {2, 7},
                 {6, 3, 2, 5, 5},
                 {4, 2, 2, 6, 2}});
  dropping = Array({5, 6, 7, 5, 3, 5});
  print_state(state);
  RET ret = solveK(state, 4);
  
  // print_state(ret.first);
  // print_state(state);
  // Act a = ret.second[0];
  // printf("\n%d->%d\n", a.first, a.second);
  // step(state, a);
  // print_state(state);
  // drop(state);
  // printf("\ndropping...\n");
  // print_state(state);
  // a = ret.second[1];
  // printf("\n%d->%d\n", a.first, a.second);
  // step(state, a);
  // print_state(state);
  // a = ret.second[2];
  // printf("\n%d->%d\n", a.first, a.second);
  // step(state, a);
  // print_state(state);
  // a = ret.second[3];
  // printf("\n%d->%d\n", a.first, a.second);
  // step(state, a);
  // print_state(state);

  print_state(ret.first);
};

int main() {
  State state;
  test3(state);
};